src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1315 @@
+	struct hash_hdl	*hdl;
+	d_list_t	*rlink;
+	mode_t		mode;
+	int		rc = 0;
+
+	/* If we aren't caching, just call dfs_lookup */
+	if (dfs_sys->dfs_hash == NULL) {
+		rc = dfs_lookup(dfs_sys->dfs, sys_path->dir_name, O_RDWR,
+				&sys_path->parent, &mode, NULL);
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "dfs_lookup() %s failed "DF_RC"\n",
+				sys_path->dir_name, DP_RC(rc));
+			return rc;
+		}
+
+		/* We only cache directories */
+		if (!S_ISDIR(mode)) {
+			dfs_release(sys_path->parent);
+			return ENOTDIR;
+		}
+		return rc;
+	}
+
+	/* If cached, return it */
+	rlink = d_hash_rec_find(dfs_sys->dfs_hash, sys_path->dir_name,
+				sys_path->dir_name_len);
+	if (rlink != NULL) {
+		hdl = hash_hdl_obj(rlink);
+		D_GOTO(out, rc = 0);
+	}
+
+	/* Not cached, so create an entry and add it */
+	D_ALLOC_PTR(hdl);
+	if (hdl == NULL)
+		return ENOMEM;
+
+	hdl->name_len = sys_path->dir_name_len;
+	D_STRNDUP(hdl->name, sys_path->dir_name, sys_path->dir_name_len);
+	if (hdl->name == NULL)
+		D_GOTO(free_hdl, ENOMEM);
+
+	/* Start with 2 so we have exactly 1 reference left
+	 * when dfs_sys_umount is called.
+	 */
+	atomic_store_relaxed(&hdl->ref, 2);
+
+	/* Lookup name in dfs */
+	rc = dfs_lookup(dfs_sys->dfs, sys_path->dir_name, O_RDWR, &hdl->obj,
+			&mode, NULL);
+	if (rc != 0) {
+		D_ERROR("dfs_lookup() %s failed "DF_RC"\n",
+			sys_path->dir_name, DP_RC(rc));
+		D_GOTO(free_hdl_name, rc);
+	}
+
+	/* We only cache directories */
+	if (!S_ISDIR(mode))
+		D_GOTO(free_hdl_obj, rc = ENOTDIR);
+
+	/* call find_insert in case another thread added the same entry
+	 * after calling find.
+	 */
+	rlink = d_hash_rec_find_insert(dfs_sys->dfs_hash, hdl->name,
+				       sys_path->dir_name_len, &hdl->entry);
+	if (rlink != &hdl->entry) {
+		/* another thread beat us. Use the existing entry. */
+		sys_path->parent = hash_hdl_obj(rlink)->obj;
+		sys_path->rlink = rlink;
+		return 0;
+	}
+
+out:
+	sys_path->parent = hdl->obj;
+	sys_path->rlink = rlink;
+	return rc;
+
+free_hdl_obj:
+	dfs_release(hdl->obj);
+free_hdl_name:
+	D_FREE(hdl->name);
+free_hdl:
+	D_FREE(hdl);
+	return rc;
+}
+
+/**
+ * Parse path into dirname and basename.
+ */
+static int parse_path(const char *path,
+		      char **_dir_name, size_t *_dir_name_len,
+		      char **_name, size_t *_name_len)
+{
+	char	*f1 = NULL;
+	char	*f2 = NULL;
+	char	*dir_name = NULL;
+	char	*name = NULL;
+	size_t	dir_name_len;
+	size_t	name_len;
+	int	rc = 0;
+
+	if (path == NULL || _name == NULL || _dir_name == NULL)
+		return EINVAL;
+	if (path[0] != '/')
+		return EINVAL;
+
+	if (strncmp(path, "/", 2) == 0) {
+		D_STRNDUP(*_dir_name, "/", 2);
+		if (*_dir_name == NULL)
+			return ENOMEM;
+		*_name = NULL;
+		*_dir_name_len = 1;
+		*_name_len = 0;
+		return 0;
+	}
+
+	D_STRNDUP(f1, path, PATH_MAX);
+	if (f1 == NULL)
+		D_GOTO(out, rc = ENOMEM);
+
+	D_STRNDUP(f2, path, PATH_MAX);
+	if (f2 == NULL)
+		D_GOTO(out, rc = ENOMEM);
+
+	dir_name = dirname(f1);
+	name = basename(f2);
+
+	dir_name_len = strnlen(dir_name, PATH_MAX);
```

Comment:
Why not do this check once on path, before copying it twice?  You could do a smaller number of checks, and before allocating resources.  Neither dirname or basename are going to return a string longer then they're passed in.

Reply:
Ah, I like that. Thanks!